### DISCUSSION TITLE
expose `bin/check*.ps1` tools in a top-level subcommand

### DISCUSSION THREAD

**Discussion by goyalyashpal** [ScoopInstaller/Scoop#5868] [Ideas]
### 1. would it make sense to expose the `bin/check*.ps1` tools in a subcommand? like `scoop dev` or smth?

these both things (local dirs as buckets (resolved in #5867), and scoop-bin utils exposed under a top-level command) would make managing local manifests much easier:

```console
$ # Imagined:
$ scoop dev checkurl mylocal/neovim-dev     # for "mylocal" local bucket
$ scoop dev checkurl ./neovim-dev           # for relative path from "." current directory
$ scoop dev checkurl d:/dir/neovim-dev      # for absolute path
...
```

currently, it requires following things:

```console
$ declare spf=".../scoop/apps/scoop/current/bin"
$ declare manif=".../manifests"

$ powershell -f $spf/checkurl.ps1 <app> $manif    
```

```console
$ # Or when trying to run all 3 checks in one call:
$ echo -n check{urls,ver,hashes} | xargs -ti -d' ' powershell -F $spf/{}.ps1 neovim-dev $manif
$ echo -n check{ver,hashes} | xargs -ti -d' ' powershell -F $spf/{}.ps1 neovim-dev $manif -up
```

```console
> # the powershell version (to run checkver on all manifests)
> ls | foreach {powershell -f $spf/checkver.ps1 $manif/$_}

> # or semi-equivalently this:
> lsd -1 | foreach {powershell -f $spf/checkver.ps1 .\$_}

> # 2024.05.29 <app> ⬅ *: apps corresponding to all the valid manifests in the bucket
> powershell -f $spf/checkver.ps1 * $local_bucket_path
```

### versions

```console
$ scoop --version | head -n2
Current Scoop version:
v0.3.1 - Released at 2022-11-15

```

**Comment by rashil2000** [comment]
This is a good idea, I've thought about this sometimes too. There was an issue/discussion regarding this, I can't seem to find it 😥

**Reply by goyalyashpal** [reply]
what more, `scoop dev` can itself operate as a function, take the 3rd word (1-indexed) as its argument, and run that as `<arg>.ps1`
that is, of course if all the tools (expected to be exposed to user) _will continue to_ follow the glob-pattern `*.ps1`

```console
$ declare spf="/d/userfiles/scoop/apps/scoop/current/bin"

$ ls $spf
auto-pr.ps1      checkver.ps1    install.ps1           scoop.ps1
checkhashes.ps1  describe.ps1    missing-checkver.ps1  test.ps1
checkurls.ps1    formatjson.ps1  refresh.ps1           uninstall.ps1
```

**Reply by goyalyashpal** [reply]
> ```console
> $ scoop dev checkurl neovim-dev
> ```

and as for this, well, if the manifests are coming from the local bucket, then that, or else the path would have to be specified. 

so, for local bucket, the feasible one would be (something like):

```console
$ # local bucket - regardless of path
$ scoop dev checkurl local/neovim-dev

$ # current directory
$ scoop dev checkurl ./neovim-dev
```

some rule will have to be specified for the case where there's a conflict, i.e. say `local` is name of both the local subdirectory, as well a local bucket. 

one such set of unambiguating rules can be:

* always prefer local bucket name in case of `bucket/app` or `app`
* use `./` for current directory relative path i.e. `./rel_dir/app` and `./app` etc
* `*:/` for absolute path i.e. `e:/abs/path/app`

_edit:_ yeah, basically same as how the bash resolves the executables from path
